include/dem/adaptive_sparse_contacts.h

```diff
@@ -0,0 +1,522 @@
+   * will wrongly propagate mobility status to the next cells.
+   *
+   * @param cell_id The current cell index
+   *
+   * @param dof_indices The vector of the DoF indices of the cell
+   *
+   * @param node_id The current node index (periodic or not)
+   *
+   * @param mobility_status The mobility status to assign
+   *
+   */
+  inline void
+  assign_mobility_status(unsigned int                          cell_id,
+                         std::vector<types::global_dof_index> &dof_indices,
+                         const int                             cell_status,
+                         const int                             node_status)
+  {
+    cell_mobility_status.insert({cell_id, cell_status});
+
+    // Assign mobility status to nodes without overwriting empty or mobile nodes
+    // in regard to the case
+    for (auto node_id : dof_indices)
+      {
+        // Prevailing mobility status of the node and assignation
+        int status_assigned = std::max(node_status, mobility_at_nodes(node_id));
+        mobility_at_nodes(node_id) = status_assigned;
+
+        // Check if node has a periodic node and assign the same mobility status
+        // if prevailing on the one on the periodic node
+        auto it = periodic_node_ids.find(node_id);
+        if (it != periodic_node_ids.end())
+          {
+            mobility_at_nodes(it->second) =
+              std::max(status_assigned, mobility_at_nodes(it->second));
+          }
+      }
+  }
+
+  /**
+   * @brief Assign the mobility status to the cell. No need to update the node
+   * status. Only use for active status assignment since nodes are not evaluated
+   * after this step.
+   *
+   * @param cell_id The current cell index
+   *
+   * @param mobility_status The mobility status to assign
+   *
+   */
+  inline void
+  assign_mobility_status(unsigned int cell_id, const int cell_status)
+  {
+    cell_mobility_status.insert({cell_id, cell_status});
+  }
+
+  // Set of locally owned and ghost cells: <local/ghost cells>
+  // Used to loop over only the locally owned and ghost cells without looping
+  // over all the cells in the triangulation numerous times.
+  std::set<typename DoFHandler<dim>::active_cell_iterator>
+    local_and_ghost_cells;
+
+  // Map of cell mobility status: <cell index: mobility status>
+  typename DEM::dem_data_structures<dim>::cell_index_int_map
+    cell_mobility_status;
+
+  // Vector of mobility status at nodes: [mobility status]
+  // Used to check the value at node to determine the mobility status of the
+  // cell, this type of vector is used to allow update values in parallel
+  LinearAlgebra::distributed::Vector<int> mobility_at_nodes;
+
+  // Map of periodic nodes: <periodic node index: coinciding node index>
+  std::unordered_map<unsigned int, unsigned int> periodic_node_ids;
+
+  // Map of cell velocities and accelerations * dt:
+  // <cell iterator: <velocity, acceleration * dt>>
+  std::map<typename Triangulation<dim>::active_cell_iterator,
```

Comment:
Same here, can you use an unordered map
No need for Ankerl shenanigans, but if you don't need the sorting and the ordering,  no need to sort

Reply:
No hash for cell iterator